File: implementations/cpp/logic_and_proof_techniques.cpp
/**
 * @file logic_and_proof_techniques.cpp
 * @brief Computational verification of logical equivalences, induction, and loop invariants.
 *
 * Implements truth-table verification of logical laws (De Morgan, contraposition),
 * automated inductive step checkers, and formally instrumented algorithms with runtime invariant assertions.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <functional>
#include <cassert>
#include <cstdint>
#include <algorithm>

namespace dsa {

/**
 * @brief Verifies if a boolean expression f(P, Q) is a tautology (True for all assignments).
 */
inline bool is_tautology_2vars(const std::function<bool(bool, bool)>& expr) {
    const bool states[2] = {false, true};
    for (bool P : states) {
        for (bool Q : states) {
            if (!expr(P, Q)) return false;
        }
    }
    return true;
}

/**
 * @brief Logical implication: P => Q is equivalent to (!P || Q).
 */
inline constexpr bool implies(bool p, bool q) {
    return !p || q;
}

/**
 * @brief Automated verification of weak mathematical induction for claim P(n) on range [base, max_n].
 * Verifies base case P(base) and inductive step P(k) => P(k+1) for all k.
 */
inline bool verify_induction(uint32_t base, uint32_t max_n, const std::function<bool(uint32_t)>& P) {
    // 1. Base case verification
    if (!P(base)) return false;

    // 2. Inductive step transmission check
    for (uint32_t k = base; k < max_n; ++k) {
        bool hyp = P(k);
        bool next = P(k + 1);
        if (!implies(hyp, next)) {
            return false;
        }
    }
    return true;
}

/**
 * @brief Binary search instrumented with explicit formal loop invariant checks.
 * Invariant: If target exists in arr, it must lie within index range [low, high].
 */
inline int binary_search_with_invariant(const std::vector<int>& arr, int target) {
    int low = 0;
    int high = static_cast<int>(arr.size()) - 1;

    // Helper lambda to check invariant
    auto check_invariant = [&](int l, int h) {
        for (size_t i = 0; i < arr.size(); ++i) {
            if (arr[i] == target) {
                // Target must be within [l, h]
                assert(static_cast<int>(i) >= l && static_cast<int>(i) <= h);
            }
        }
    };

    // Initialization check
    check_invariant(low, high);

    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
        // Maintenance check
        check_invariant(low, high);
    }

    // Termination check: Target not found, meaning it cannot exist in array
    for (int val : arr) {
        assert(val != target);
    }
    return -1;
}

/**
 * @brief Dutch National Flag 3-way partition instrumented with loop invariant.
 * Invariant partitions array into:
 *   [0 ... low-1]: strictly 0
 *   [low ... mid-1]: strictly 1
 *   [mid ... high]: unclassified
 *   [high+1 ... n-1]: strictly 2
 */
inline void dutch_national_flag_with_invariant(std::vector<int>& nums) {
    int low = 0;
    int mid = 0;
    int high = static_cast<int>(nums.size()) - 1;

    auto verify_dutch_invariant = [&](int l, int m, int h) {
        for (int i = 0; i < l; ++i) assert(nums[i] == 0);
        for (int i = l; i < m; ++i) assert(nums[i] == 1);
        for (size_t i = static_cast<size_t>(h + 1); i < nums.size(); ++i) assert(nums[i] == 2);
    };

    // Initialization
    verify_dutch_invariant(low, mid, high);

    while (mid <= high) {
        if (nums[mid] == 0) {
            std::swap(nums[low], nums[mid]);
            ++low;
            ++mid;
        } else if (nums[mid] == 1) {
            ++mid;
        } else {
            std::swap(nums[mid], nums[high]);
            --high;
        }
        // Maintenance
        verify_dutch_invariant(low, mid, high);
    }

    // Termination: unclassified interval [mid, high] is empty (mid > high)
    verify_dutch_invariant(low, mid, high);
}

} // namespace dsa

int main() {
    std::cout << "Running Logic and Proof Techniques C++17 unit tests..." << std::endl;

    // Test 1: De Morgan's Laws equivalence verification
    // !(P && Q) <=> (!P || !Q)
    assert(dsa::is_tautology_2vars([](bool p, bool q) {
        return (!(p && q)) == (!p || !q);
    }));

    // !(P || Q) <=> (!P && !Q)
    assert(dsa::is_tautology_2vars([](bool p, bool q) {
        return (!(p || q)) == (!p && !q);
    }));

    // Test 2: Contrapositive Equivalence
    // (P => Q) <=> (!Q => !P)
    assert(dsa::is_tautology_2vars([](bool p, bool q) {
        return dsa::implies(p, q) == dsa::implies(!q, !p);
    }));

    // Test 3: Implication identity
    // (P => Q) <=> (!P || Q)
    assert(dsa::is_tautology_2vars([](bool p, bool q) {
        return dsa::implies(p, q) == (!p || q);
    }));

    // Test 4: Verification of induction: Sum of first n integers claim P(n): sum_{i=1}^n i == n(n+1)/2
    assert(dsa::verify_induction(1, 1000, [](uint32_t n) {
        uint64_t sum = 0;
        for (uint32_t i = 1; i <= n; ++i) sum += i;
        uint64_t formula = static_cast<uint64_t>(n) * (n + 1) / 2;
        return sum == formula;
    }));

    // Test 5: Verification of induction: Sum of squares P(n): sum i^2 == n(n+1)(2n+1)/6
    assert(dsa::verify_induction(1, 500, [](uint32_t n) {
        uint64_t sum = 0;
        for (uint32_t i = 1; i <= n; ++i) sum += static_cast<uint64_t>(i) * i;
        uint64_t formula = static_cast<uint64_t>(n) * (n + 1) * (2 * n + 1) / 6;
        return sum == formula;
    }));

    // Test 6: Binary search loop invariant execution
    std::vector<int> sorted_arr = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};
    assert(dsa::binary_search_with_invariant(sorted_arr, 23) == 5);
    assert(dsa::binary_search_with_invariant(sorted_arr, 2) == 0);
    assert(dsa::binary_search_with_invariant(sorted_arr, 91) == 9);
    assert(dsa::binary_search_with_invariant(sorted_arr, 40) == -1);

    // Test 7: Dutch National Flag loop invariant execution
    std::vector<int> colors = {2, 0, 2, 1, 1, 0, 2, 1, 0, 0, 1, 2};
    dsa::dutch_national_flag_with_invariant(colors);
    assert(std::is_sorted(colors.begin(), colors.end()));

    std::cout << "[PASS] All Logic and Proof Techniques C++ unit tests passed." << std::endl;
    return 0;
}

File: implementations/python/logic_and_proof_techniques.py
"""
Formal Logic, Mathematical Induction, and Loop Invariant Verification.

Provides truth-table verification of logical equivalences (De Morgan, contraposition),
inductive proof checkers, and loop-invariant-instrumented algorithmic procedures.
"""

import unittest
from typing import Callable, List


def implies(p: bool, q: bool) -> bool:
    """Logical implication: P => Q."""
    return (not p) or q


def is_tautology_2vars(expr: Callable[[bool, bool], bool]) -> bool:
    """Checks whether boolean expression f(P, Q) is true for all 4 truth assignments."""
    for p in (False, True):
        for q in (False, True):
            if not expr(p, q):
                return False
    return True


def verify_induction(base: int, max_n: int, prop: Callable[[int], bool]) -> bool:
    """
    Empirically verifies weak induction for predicate prop(n) on [base, max_n].
    Checks base case and transmission implication prop(k) => prop(k+1).
    """
    if not prop(base):
        return False
    for k in range(base, max_n):
        if not implies(prop(k), prop(k + 1)):
            return False
    return True


def binary_search_with_invariant(arr: List[int], target: int) -> int:
    """
    Binary search formally instrumented with loop invariant assertions.
    Invariant: If target is in arr, it must be within arr[low .. high].
    """
    low = 0
    high = len(arr) - 1

    def assert_invariant(l: int, h: int):
        for idx, val in enumerate(arr):
            if val == target:
                assert l <= idx <= h, f"Invariant violated: idx={idx} not in [{l}, {h}]"

    assert_invariant(low, high)

    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
        assert_invariant(low, high)

    assert target not in arr
    return -1


def dutch_national_flag_with_invariant(nums: List[int]) -> None:
    """
    Sorts an array of 0s, 1s, and 2s in-place with invariant checking.
    Invariant:
      nums[0 .. low-1] == 0
      nums[low .. mid-1] == 1
      nums[mid .. high] == unclassified
      nums[high+1 .. end] == 2
    """
    low = 0
    mid = 0
    high = len(nums) - 1

    def assert_invariant(l: int, m: int, h: int):
        for i in range(l):
            assert nums[i] == 0
        for i in range(l, m):
            assert nums[i] == 1
        for i in range(h + 1, len(nums)):
            assert nums[i] == 2

    assert_invariant(low, mid, high)

    while mid <= high:
        if nums[mid] == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            mid += 1
        else:
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1
        assert_invariant(low, mid, high)

    assert_invariant(low, mid, high)


class TestLogicAndProofTechniques(unittest.TestCase):
    def test_de_morgan_laws(self):
        # !(P && Q) <=> (!P || !Q)
        self.assertTrue(is_tautology_2vars(lambda p, q: (not (p and q)) == ((not p) or (not q))))
        # !(P || Q) <=> (!P && !Q)
        self.assertTrue(is_tautology_2vars(lambda p, q: (not (p or q)) == ((not p) and (not q))))

    def test_contrapositive_equivalence(self):
        # (P => Q) <=> (!Q => !P)
        self.assertTrue(is_tautology_2vars(lambda p, q: implies(p, q) == implies(not q, not p)))

    def test_implication_identity(self):
        # (P => Q) <=> (!P || Q)
        self.assertTrue(is_tautology_2vars(lambda p, q: implies(p, q) == ((not p) or q)))

    def test_induction_sum(self):
        # sum_{i=1}^n i == n(n+1)/2
        self.assertTrue(
            verify_induction(
                1, 500, lambda n: sum(range(1, n + 1)) == (n * (n + 1)) // 2
            )
        )

    def test_binary_search_invariant(self):
        arr = [1, 3, 5, 7, 9, 11, 13, 15]
        self.assertEqual(binary_search_with_invariant(arr, 7), 3)
        self.assertEqual(binary_search_with_invariant(arr, 1), 0)
        self.assertEqual(binary_search_with_invariant(arr, 15), 7)
        self.assertEqual(binary_search_with_invariant(arr, 8), -1)

    def test_dutch_national_flag_invariant(self):
        nums = [2, 0, 2, 1, 1, 0, 2, 1, 0]
        dutch_national_flag_with_invariant(nums)
        self.assertEqual(nums, sorted([2, 0, 2, 1, 1, 0, 2, 1, 0]))


if __name__ == "__main__":
    unittest.main()
